app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,236 @@
+    end
+
+    # Checks the cache for an unexpired token and returns that.
+    # If the cache has expired, retrieves a new token and returns it
+    # @return [String] Auth token
+    def token
+      cache_value = Rails.cache.read(cache_key)
+      return cache_value unless IdentityConfig.store.arcgis_token_sync_request_enabled
+      if cache_value.nil?
+        retrieve_token&.fetch(:token)
+      else
+        handle_expired_token(cache_value)&.fetch(:token)
+      end
+    end
+
+    # @param [Object] cache_value the entry to save in cache
+    # @param [Number] expires_at the unix time cache entry should expire
+    def save_token(cache_value, expires_at)
+      Rails.cache.write(cache_key, cache_value, expires_at: expires_at)
+      # If using a redis cache we have to manually set the expires_at. This is because we aren't
+      # using a dedicated Redis cache and instead are just using our existing Redis server with
+      # mixed usage patterns. Without this cache entries don't expire.
+      # More at https://api.rubyonrails.org/classes/ActiveSupport/Cache/RedisCacheStore.html
+      Rails.cache.try(:redis)&.expireat(cache_key, expires_at.to_i)
+    end
+
+    # Makes a request to retrieve a new token
+    # it expires after 1 hour
+    # @return [String] Auth token
+    def retrieve_token
+      token, expires = request_token.fetch_values('token', 'expires')
+      expires_at = Time.zone.at(expires / 1000).to_f
+      cache_entry = {
+        token: token,
+        expires_at: expires_at,
+      }
+
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled == true
+        cache_entry = cache_entry.merge(
+          { sliding_expires_at: prefetch_ttl >= 0 ? expires_at - 3 * prefetch_ttl : expires_at },
+        )
+      end
+      save_token(cache_entry, expires_at)
+      cache_entry
+    end
+
+    def remove_token
+      Rails.cache.delete(cache_key)
+    end
+
+    private
+
+    def handle_expired_token(cache_value)
+      cache_value = wrap_raw_token(cache_value) unless cache_value.is_a?(Hash)
+      expires_at = cache_value.fetch(:expires_at, nil)
+      sliding_expires_at = cache_value&.fetch(:sliding_expires_at, nil)
+      # start fetch new token before actual expires time
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled
+        now = Time.zone.now.to_f
+        if sliding_expires_at && now >= sliding_expires_at
+          # if passed sliding expiration time(exp-3*prefetch_ttl)
+          # but not 2*prefetch_ttl from hard expiration time
+          if now > (sliding_expires_at + prefetch_ttl)
+            cache_value = retrieve_token
+          else
+            # extend sliding expiration time
+            cache_value[:sliding_expires_at] = sliding_expires_at + prefetch_ttl
+            save_token(cache_value, expires_at)
+          end
+        end
+      elsif expires_at && expires_at <= now
+        cache_value = retrieve_token
+      end
+      cache_value
+    end
```

Comment:
@NavaTim played with it a little bit, got kind of circular dependency with strategy pattern, trying subclassing now.

Reply:
> Act as an escape route hopefully never needed?

I think it's acceptable (though undesirable) to use this as backup to the job, but that does increase the risk of excessive authentication calls. It may be better to turn it off entirely, but for now we will need both.

> @NavaTim played with it a little bit, got kind of circular dependency with strategy pattern, trying subclassing now.

I think the sliding window approach used here may be more complicated than necessary. We can use a specific amount of time before expiration instead, and maybe that will make this logic simpler.